src/heuristic.c

```diff
@@ -0,0 +1,221 @@
+        float *nir = malloc(img_buffer[imgnum].x_size * img_buffer[imgnum].y_size * sizeof(float));
+        float *swir = malloc(img_buffer[imgnum].x_size * img_buffer[imgnum].y_size * sizeof(float));
+        if (b == NULL || r == NULL || nir == NULL || swir == NULL){
+                printf("Unable to allocate memory for band array");
+                return 1;
+        }
+
+        //Reading the BSQ formatted raster file. Populating b, r, nir, swir band arrays
+        if (img_buffer[0].in_interleaving == BSQ) {
+                for (int j = 0; j < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; j++) {
+                        fscanf(fptr, "%f ", b+j);
+                }
+
+                for (int j = 0; j < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; j++) {
+                        fscanf(fptr, "%f ", r+j);
+                }
+ 
+                for (int j = 0; j < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; j++) {
+                        fscanf(fptr, "%f ", nir+j);
+                }
+
+                for (int j = 0; j < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; j++) {
+                        fscanf(fptr, "%f ", swir+j);
+                }
+
+                //Loop through each index of the band arrays and running isCloud() on each iteration (each pixel)
+                for (int i = 0; i < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; i++) {
+                        printf("b: %f\nr: %f\nnir: %f\nswir: %f\n\n", *(b+i), *(r+i), *(nir+i), *(swir+i));
+                        if (isCloud(*(b+i), *(r+i), *(nir+i), *(swir+i))) {
+                                raw_heuristic+=1;
+                        }
+                }
+                
+                //Free allocated memory of the band arrays
+                free(b);        
+                free(r);
+                free(nir);
+                free(swir);
+                fclose(fptr);
+
+                //Returns number of cloudy pixels / all pixels.
+                return raw_heuristic / (img_buffer[imgnum].x_size * img_buffer[imgnum].y_size);
+
+        //Reading the BIP formatted raster file. Populating b, r, nir, swir band arrays
+        } else if (img_buffer[0].in_interleaving == BIP) {
+                for (int i = 0; i < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; i++) {
+                        fscanf(fptr, "%f %f %f %f ", b+i, r+i, nir+i, swir+i);
+                }
+
+                for (int i = 0; i < img_buffer[imgnum].x_size * img_buffer[imgnum].y_size; i++) {
+                        printf("b: %f\nr: %f\nnir: %f\nswir: %f\n\n", *(b+i), *(r+i), *(nir+i), *(swir+i));
+                        if(isCloud(*(b+i), *(r+i), *(nir+i), *(swir+i))){
+                                raw_heuristic+=1;
+                        }
+                }
+
+                free(b);        
+                free(r);
+                free(nir);
+                free(swir);
```

Comment:
Same here, theres some companion to `pvPortMalloc`, I think it's `vPortFree` to be thread safe.